### PR TITLE
client: impl intercept syscall

### PR DESCRIPTION
**Pull request by Tricky511**
**client side**:
- [x] open
- [ ] rename
- [x] truncate
- [x] mkdir
- [x] rmdir
- [x] getdents64
- [x] unlink
- [x] stat
- [ ] pread, preadv
- [ ] pwrite, pwritev

### CODE REVIEW

File: src/common/serialization.rs [hunk 1]
```diff
@@ -146,6 +156,83 @@ impl FileAttrSimple {
             blksize: 0,
         }
     }
+
+    pub fn tostat(&self, statbuf: &mut [u8]) {
```
Comment (luanshaotong) [inline src/common/serialization.rs:160]: How about to convert `FileAttrSimple` into `stat` in an `into()` function and just convert `stat` into `[u8]` in this function?

File: src/server/distributed_engine.rs [hunk 2]
```diff
@@ -6,6 +6,7 @@ use crate::common::{
 };
 
 use crate::rpc::client::Client;
+use libc::off_t;
```
Comment (luanshaotong) [inline src/server/distributed_engine.rs:9]: Do not use libc types on server. The type converting should only occur on client.

File: src/server/storage_engine/default_engine.rs [hunk 3]
```diff
@@ -8,6 +8,7 @@ use crate::server::distributed_engine::path_split;
 use super::EngineError;
 
 use super::StorageEngine;
+use libc::off_t;
```
Comment (luanshaotong) [inline src/server/storage_engine/default_engine.rs:11]: The same as above.

File: src/server/storage_engine/mod.rs [hunk 4]
```diff
@@ -2,6 +2,7 @@
 //
 // SPDX-License-Identifier: Apache-2.0
 
+use libc::off_t;
```
Comment (luanshaotong) [inline src/server/storage_engine/mod.rs:5]: The same as above.

File: src/common/serialization.rs (hunk 1 above)
Comment (Tricky511) [reply src/common/serialization.rs:160]: I think this reduces one copy

Comment (luanshaotong) [reply src/common/serialization.rs:160]: You are right.

File: src/server/distributed_engine.rs (hunk 2 above)
Comment (Tricky511) [reply src/server/distributed_engine.rs]: done

File: src/server/storage_engine/default_engine.rs (hunk 3 above)
Comment (Tricky511) [reply src/server/storage_engine/default_engine.rs]: done

File: src/server/storage_engine/mod.rs (hunk 4 above)
Comment (Tricky511) [reply src/server/storage_engine/mod.rs]: done

File: intercept/src/lib.rs [hunk 5]
```diff
@@ -45,617 +49,927 @@ extern "C" fn initialize() {
...
+                Ok(_) => *result = 0,
+                Err(e) => {
+                    *result = -e as isize;
+                }
+            }
+            InterceptResult::Hook
         }
-        _ => 1,
+        // int statfs(const char *path, struct statfs *buf);
+        // SYS_statfs => {
```
Comment (luanshaotong) [inline intercept/src/lib.rs:952]: Are those comments necessary?

File: intercept/build.rs [hunk 6]
```diff
@@ -1,8 +1,7 @@
...
-// SPDX-License-Identifier: Apache-2.0
+fn main() {
+    let dst = cmake::build("../syscall_intercept");
 
-fn main() -> Result<(), Box<dyn std::error::Error>> {
-    println!(r"cargo:rustc-link-search=native=./lib");
-    Ok(())
+    println!("cargo:rustc-link-search=native={}/lib", dst.display());
+    println!("cargo:rustc-link-lib=static=syscall_intercept");
+    println!("cargo:rustc-link-lib=capstone");
```
[the remark this answers is not in the document]
Comment (Tricky511) [reply intercept/build.rs:6]: This is not a log, this is building the required dependency library

File: intercept/src/lib.rs (hunk 5 above)
Comment (Tricky511) [reply intercept/src/lib.rs]: deleted

File: intercept/src/client.rs [hunk 7]
```diff
@@ -14,7 +17,8 @@ pub struct Client {
     pub inodes_reverse: DashMap<u64, String>,
     handle: tokio::runtime::Handle,
 }
-
+const CHUNK_SIZE: i64 = 65535;
```
[the remark this answers is not in the document]
Comment (Tricky511) [reply intercept/src/client.rs]: done

**Review by luanshaotong — APPROVED** [review]
LGTM

**Pull request merged**